pkg/pipeline/stream/source_test.go

```diff
@@ -0,0 +1,174 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stream
+
+import (
+	"context"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/conduitio/conduit/pkg/connector/mock"
+	"github.com/conduitio/conduit/pkg/foundation/metrics/noop"
+	"github.com/conduitio/conduit/pkg/plugin"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/golang/mock/gomock"
+	"github.com/google/uuid"
+	"github.com/matryer/is"
+)
+
+func TestSourceNode_Run(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+
+	src, wantRecords := newMockSource(ctrl, 10, nil)
+	node := &SourceNode{
+		Name:          "source-node",
+		Source:        src,
+		PipelineTimer: noop.Timer{},
+	}
+	out := node.Pub()
+
+	nodeDone := make(chan struct{})
+	go func() {
+		defer close(nodeDone)
+		err := node.Run(ctx)
+		is.NoErr(err)
+	}()
+
+	for _, wantRecord := range wantRecords {
+		gotMsg := <-out
```

Comment:
How likely is it that we wait for the message indefinitely (due to a bug in SourceNode)? I'm wondering if it would be an overkill to have a timeout here.

Reply:
I took the opportunity to add utilities for receiving from a channel with a timeout and for waiting for a waitgroup with a timeout (remember this? https://github.com/ConduitIO/conduit-connector-sdk/commit/a9290b473be448836a4a0225afffa1215eda8307). Then I added the timeout here and simplified test code that was using the same functionality.

Scope creep? Nooo, just "leaving the campsite nicer than I found it" 😇